packages/web3/src/address/address.test.ts

```diff
@@ -17,194 +17,316 @@ along with the library. If not, see <http://www.gnu.org/licenses/>.
-    expect(validateAddress('yya86C6UemCeLs5Ztwjcf2Mp2Kkt4mwzzRpBiG6qQ9kj')).toBeUndefined()
-    expect(() => validateAddress('yya86C6UemCeLs5Ztwjcf2Mp2Kkt4mwzzRpBiG6qQ9k')).toThrow('Invalid address:')
-    expect(() => validateAddress('asd')).toThrow('Invalid multisig address: asd')
-    expect(() => validateAddress('asdasdf')).toThrow('Invalid multisig address')
-    expect(() =>
-      validateAddress('2jW1n2icPtc55Cdm8TF9FjGH681cWthsaZW3gaUFekFZepJoeyY3ZbY7y5SCtAjyCjLL24c4L2Vnfv3KDdAypCddfAY1')
-    ).toThrow('Invalid address:')
-    // both n and m are 0
-    expect(() => validateAddress('LUw')).toThrow('Invalid multisig address')
-    expect(() =>
-      validateAddress('2jVWAcAPphJ8ueZNG1BPwbfPFjjbvorprceuqzgmJQ1ZRyELRpWgARvdB3T9trqpiJs7f4GkudPt6rQLnGbQYqq2NCi')
-    ).toThrow('Invalid multisig address, n: 2, m: 3')
-    expect(() => validateAddress('thebear')).toThrow('Invalid multisig address')
+import { LockupScript } from '../codec/lockup-script-codec'
+
+describe('address', () => {
+  describe('validateAddress', () => {
+    it('should validate valid P2PKH addresses', () => {
+      expect(() => validateAddress('15EM5rGtt7dPRZScE4Z9oL2EDfj84JnoSgq3NNgdcGFyu')).not.toThrow()
+      expect(() => validateAddress('1D59jXR9NpD9ZQqZTRVcVbKVh6ko5TUMt89WvkA8P9P7w')).not.toThrow()
+    })
+
+    it('should validate valid P2MPKH addresses', () => {
+      expect(() =>
+        validateAddress(`'2jW1n2icPtc55Cdm8TF9FjGH681cWthsaZW3gaUFekFZepJoeyY3ZbY7y5SCtAjyCjLL24c4L2Vnfv3KDdAypCddfAY'`)
+      ).not.toThrow()
+    })
+
+    it('should validate valid P2SH addresses', () => {
+      expect(() => validateAddress('eBrjfQNeyUCuxE4zpbfMZcbS3PuvbMJDQBCyk4HRHtX4')).not.toThrow()
+    })
+
+    it('should validate valid P2C addresses', () => {
+      expect(() => validateAddress('yya86C6UemCeLs5Ztwjcf2Mp2Kkt4mwzzRpBiG6qQ9kj')).not.toThrow()
+    })
+
+    it('should throw error for empty address', () => {
+      expect(() => validateAddress('')).toThrow('Address is empty')
+    })
+
+    it('should throw error for invalid base58 string', () => {
+      expect(() => validateAddress('InvalidBase58!!')).toThrow('Non-base58 character')
+    })
+
+    it('should throw error for invalid address type', () => {
+      const invalidAddressType = 0x04
+      const hash = randomBytes(32)
+      const bytes = Buffer.concat([Buffer.from([invalidAddressType]), hash])
+      const address = bs58.encode(bytes)
+      expect(() => validateAddress(address)).toThrow(`Invalid address: ${address}`)
+    })
+
+    it('should throw error for invalid multisig address', () => {
+      const invalidAddressType = 0x01
+      const invalidData = randomBytes(10)
+      const bytes = Buffer.concat([Buffer.from([invalidAddressType]), invalidData])
+      const address = bs58.encode(bytes)
+      expect(() => validateAddress(address)).toThrow(`Invalid multisig address: ${address}`)
```

Comment:
The exception is not deterministic, sometimes it could return `"Invalid multisig address, n: 0, m: xxxxx"`